### PR TITLE
Initial OpenACC support

### PR DESCRIPTION
**Pull request by dalg24**
Revisited #5203 

@seyonglee please review

### CODE REVIEW

File: core/src/Kokkos_OpenACC.hpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#include <string>
+
+namespace Kokkos::Experimental::Impl {
+class OpenACCInternal;
+}
+
+namespace Kokkos::Experimental {
+
+class OpenACC {
+  Impl::OpenACCInternal* m_space_instance = nullptr;
```
Comment (crtrott) [inline core/src/Kokkos_OpenACC.hpp:77]: This should be Impl::HostSharedPtr<Impl::OpenACCInternal> Or at least make a comment that you need that later.

**Review by crtrott — CHANGES_REQUESTED** [review]

File: core/src/Kokkos_OpenACC.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/Kokkos_OpenACC.hpp:77]: Otherwise I like it.

Comment (dalg24) [reply core/src/Kokkos_OpenACC.hpp:77]: I meant to do it when we implement multiple instances support

**Comment by seyonglee** [comment]
It seems that macro statement to conditionally include the OpenACC backend (`#if defined(KOKKOS_ENABLE_OPENACC)` ... `#endif` in `Kokkos_OpenACC.h` and `Kokkos_OpenACCSpace.h`) is removed, which will cause compilation errors when targeting non-OpenACC backends, since these header files include OpenACC APIs.

File: core/src/Kokkos_OpenACC.hpp (hunk 1 above)
Comment (seyonglee) [reply core/src/Kokkos_OpenACC.hpp:77]: Do we want to change this now? or in the later PR?

**Comment by dalg24** [comment]
> It seems that macro statement to conditionally include the OpenACC backend (`#if defined(KOKKOS_ENABLE_OPENACC)` ... `#endif` in `Kokkos_OpenACC.h` and `Kokkos_OpenACCSpace.h`) is removed, which will cause compilation errors when targeting non-OpenACC backends, since these header files include OpenACC APIs.

Yes I will move the file to the `OpenACC` subdirectory which will fix the CI errors.

**Comment by seyonglee** [comment]
Other than the macro statement issue, it looks good to me too.

**Review by crtrott — APPROVED** [review]

File: core/src/OpenACC/Kokkos_OpenACC_Instance.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    std::string const& name) const {
+  Kokkos::Tools::Experimental::Impl::profile_fence_event<
+      Kokkos::Experimental::OpenACC>(
+      name,
+      Kokkos::Tools::Experimental::Impl::DirectFenceIDHandle{instance_id()},
+      [&]() {
+        //[DEBUG] disabled due to synchronous behaviors of the current
+        // parallel construct implementations. acc_wait_all();
+      });
+}
```
Comment (JBludau) [inline core/src/OpenACC/Kokkos_OpenACC_Instance.cpp:86]: I personally do not like empty functions for debugging/development, especially if they do not even have side effects. Why not have a fence here, even though it is not necessary at the current state? A `FIXME_OPENACC` would not hurt I guess

**Review by JBludau — APPROVED** [review]
No hard feelings concerning my comments, just suggestions. Looks good

File: cmake/kokkos_arch.cmake [hunk 3]
```diff
@@ -124,6 +124,13 @@ IF(KOKKOS_ENABLE_COMPILER_WARNINGS)
     LIST(REMOVE_ITEM COMMON_WARNINGS "-Wsign-compare")
   ENDIF()
 
+  # NVHPC compiler does not support -Wtype-limits.
+  IF(KOKKOS_ENABLE_OPENACC)
+    IF(KOKKOS_CXX_COMPILER_ID STREQUAL NVHPC)
+      LIST(REMOVE_ITEM COMMON_WARNINGS "-Wtype-limits")
+    ENDIF()
+  ENDIF()
```
[the remark this answers is not in the document]
Comment (dalg24) [reply cmake/kokkos_arch.cmake:132]: Dunno

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

File: cmake/kokkos_arch.cmake [hunk 4]
```diff
@@ -490,8 +497,8 @@ FUNCTION(CHECK_CUDA_ARCH ARCH FLAG)
       MESSAGE(FATAL_ERROR "Multiple GPU architectures given! Already have ${CUDA_ARCH_ALREADY_SPECIFIED}, but trying to add ${ARCH}. If you are re-running CMake, try clearing the cache and running again.")
     ENDIF()
     SET(CUDA_ARCH_ALREADY_SPECIFIED ${ARCH} PARENT_SCOPE)
-    IF (NOT KOKKOS_ENABLE_CUDA AND NOT KOKKOS_ENABLE_OPENMPTARGET AND NOT KOKKOS_ENABLE_SYCL)
-      MESSAGE(WARNING "Given CUDA arch ${ARCH}, but Kokkos_ENABLE_CUDA and Kokkos_ENABLE_OPENMPTARGET are OFF. Option will be ignored.")
+    IF (NOT KOKKOS_ENABLE_CUDA AND NOT KOKKOS_ENABLE_OPENMPTARGET AND NOT KOKKOS_ENABLE_SYCL AND NOT KOKKOS_ENABLE_OPENACC)
+      MESSAGE(WARNING "Given CUDA arch ${ARCH}, but Kokkos_ENABLE_CUDA, Kokkos_ENABLE_OPENACC, and Kokkos_ENABLE_OPENMPTARGET are OFF. Option will be ignored.")
```
[the remark this answers is not in the document]
Comment (dalg24) [reply cmake/kokkos_arch.cmake:501]: I agree with the suggestion but am tempted to ignore for the sake of time.  You can push to my branch if you want it now.

[the remark this answers is not in the document]
Comment (JBludau) [reply cmake/kokkos_arch.cmake:501]: no, we can exclude it and just do it in another pr ... it is not OpenACC related anyways

**Pull request merged**